baseplate/core.py

```diff
@@ -123,6 +124,123 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
+    the incoming requests.
+
+    :param baseplate.secrets.SecretsStore secrets_store: the application's
+        defined secrets store, where application secret lookups should be made
+    """
+    def __init__(self, secrets_store=None):
+        self.secrets = secrets_store
+
+    def make_context(self, token):
+        """:py:class:`AuthenticationContext` builder
+
+        :param token: JWT value originating from the Authentication service
+            either directly or from an upstream service
+        :rtype: :py:class:`AuthenticationContext`
+        """
+        return AuthenticationContext(token=token, secrets=self.secrets)
+
+
+class AuthenticationContext(object):
+    """Wrapper for the contextual authentication information
+
+    :param str token: the JWT value for the currently propagated authentication
+        context
+    :param baseplate.secrets.SecretsStore secrets_store: the application's
+        defined secrets store, where application secret lookups should be made
+    """
+    def __init__(self, token=None, secrets=None):
+        self.token = token
+        self.secrets = secrets
+        self.payload = {}
+        self._valid = None
+
+    def _secret(self):
+        if not self.secrets:
+            raise UndefinedSecretsException
+
+        return self.secrets.get_simple("jwt/authentication/secret")
+
+    @property
+    def valid(self):
+        """Validity of the current authentication context token
+
+        :type: bool
+        :raises: :py:class:`UndefinedSecretsException` if the
+            :py:class:`SecretsStore` has not been bound to the context handling
+            class (means that the authentication payload could not be decrypted
+            and validated)
+
+        """
+        if self._valid is not None:
+            return self._valid
+        elif self.token is None:
+            return None
+
+        try:
+            self.payload = jwt.decode(self.token, self._secret(),
+                                      algorithms='RS256')
+            self._valid = True
+        except jwt.ExpiredSignatureError:
+            self._valid = False
+        except jwt.DecodeError:
+            self._valid = False
+
+        return self._valid
+
+    @property
+    def account_id(self):
+        """Authenticated account_id for the current authenticated context
+
+        :type: account_id string or None if context authentication is invalid
+        :raises: :py:class:`UndefinedSecretsException` if the
+            :py:class:`SecretsStore` has not been bound to the context handling
+            class (means that the authentication payload could not be decrypted
+            and the user_id returned)
+        :raises: :py:class:`UndefinedAuthenticationError` if there was no
+            authentication token defined for the current context
+
+        """
+        if self.token is None:
+            raise UndefinedAuthenticationError
+        if not self.valid:
+            return None
```

Comment:
Yeah, I am not super familiar with how much we would dislike it if new fields were being added to the token and applications would need to stall for smart wrappers to access the new fields.  I can see doing what you had but not raise the `UndefinedAuthenticationError` to at least encapsulate the validation check but not attempt to do any more, it would also enforce the `payload` property to be read only, which is a plus.